File: packages/frontend-web/src/lib/api/compliance-exports.ts
import { useQuery } from '@tanstack/react-query'
import { apiClient } from '@/lib/api-client'

export type HazardExportRow = {
  deliveryNumber?: string | null
  deliveryDate?: string | null
  supplierName?: string | null
  customerId?: string | null
  article?: string | null
  bvlZulassungsnummer?: string | null
  hazardHinweise?: string | null
  sdsReference?: string | null
  sachkundeStatus?: string | null
  sdsMitgeliefert?: string | null
  adrPunkte?: number
  compliant?: boolean
}

export type HazardExportResponse = {
  year?: number | null
  rows: HazardExportRow[]
  total: number
  generated_at: string
}

export type NutrientMonth = {
  deliveries: number
  n_kg: number
  p2o5_kg: number
}

export type NutrientExportResponse = {
  year: number
  deliveryCount: number
  totalNutrientNKg: number
  totalNutrientP2o5Kg: number
  byMonth: Record<string, NutrientMonth>
  generated_at: string
}

const EMPTY_HAZARD_EXPORT: HazardExportResponse = {
  year: null,
  rows: [],
  total: 0,
  generated_at: '',
}

const EMPTY_NUTRIENT_EXPORT: NutrientExportResponse = {
  year: 0,
  deliveryCount: 0,
  totalNutrientNKg: 0,
  totalNutrientP2o5Kg: 0,
  byMonth: {},
  generated_at: '',
}

export type LotTraceResponse = {
  lot?: {
    id: string
    lotId?: string | null
    article?: string | null
    articleId?: string | null
    quantity?: number
    location?: string | null
    status?: string | null
    qualityStatus?: string | null
    origin?: string | null
  }
  events?: Array<{ type: string; date?: string | null; note?: string | null }>
  linkedDeliveries?: Array<{
    deliveryNumber?: string | null
    deliveryDate?: string | null
    article?: string | null
    quantity?: number
    customerId?: string | null
  }>
  deliveryCount?: number
  generated_at?: string
  error?: string
  lot_id?: string
}

export function useGefahrstoffExport(year?: number): ReturnType<typeof useQuery<HazardExportResponse>> {
  return useQuery({
    queryKey: ['compliance', 'exports', 'gefahrstoffdoku', year ?? null],
    queryFn: async () => {
      const params = new URLSearchParams()
      if (year) params.set('year', String(year))
      const path = params.size > 0 ? `/api/v1/compliance/exports/gefahrstoffdoku?${params.toString()}` : '/api/v1/compliance/exports/gefahrstoffdoku'
      return (await apiClient.get<HazardExportResponse>(path)).data
    },
    initialData: EMPTY_HAZARD_EXPORT,
    staleTime: 60 * 1000,
  })
}

export function useNaehrstoffExport(year: number): ReturnType<typeof useQuery<NutrientExportResponse>> {
  return useQuery({
    queryKey: ['compliance', 'exports', 'naehrstoffstrom', year],
    queryFn: async () => {
      return (await apiClient.get<NutrientExportResponse>(`/api/v1/compliance/exports/naehrstoffstrom?year=${encodeURIComponent(String(year))}`)).data
    },
    enabled: Number.isInteger(year) && year >= 2000 && year <= 2100,
    initialData: EMPTY_NUTRIENT_EXPORT,
    staleTime: 60 * 1000,
  })
}

export async function fetchLotTrace(lotId: string): Promise<LotTraceResponse> {
  const id = lotId.trim()
  if (!id) {
    throw new Error('lot_id ist erforderlich')
  }
  return (await apiClient.get<LotTraceResponse>(`/api/v1/compliance/exports/chargen-trace/${encodeURIComponent(id)}`)).data
}

export async function downloadComplianceCsv(kind: 'gefahrstoffdoku' | 'naehrstoffstrom', year?: number): Promise<void> {
  const params = new URLSearchParams()
  if (year) params.set('year', String(year))
  const endpoint = kind === 'gefahrstoffdoku' ? 'gefahrstoffdoku.csv' : 'naehrstoffstrom.csv'
  const query = params.size > 0 ? `?${params.toString()}` : ''
  const response = await apiClient.get<Blob>(`/api/v1/compliance/exports/${endpoint}${query}`, {
    responseType: 'blob',
  })
  const blob = new Blob([response.data], { type: 'text/csv;charset=utf-8' })
  const url = URL.createObjectURL(blob)
  const anchor = document.createElement('a')
  anchor.href = url
  anchor.download = kind === 'gefahrstoffdoku'
    ? (year ? `gefahrstoffdoku-${year}.csv` : 'gefahrstoffdoku.csv')
    : `naehrstoffstrom-${year ?? 'export'}.csv`
  document.body.appendChild(anchor)
  anchor.click()
  document.body.removeChild(anchor)
  URL.revokeObjectURL(url)
}

File: app/core/compliance_exports.py
from __future__ import annotations

from collections import defaultdict
from typing import Optional


def extract_hazard_export_rows(deliveries: list[dict], year: Optional[int] = None) -> list[dict]:
    rows: list[dict] = []
    for d in deliveries:
        date_str = str(d.get("date") or "")
        if year and not date_str.startswith(f"{year}-"):
            continue
        doc_compliance = d.get("psmCompliance") or {}
        for line in d.get("lines") or []:
            has_psm = any([line.get("bvlZulassungsnummer"), line.get("hazardHinweise"), line.get("sdsReference")])
            if not has_psm:
                continue
            rows.append(
                {
                    "deliveryNumber": d.get("number") or d.get("id"),
                    "deliveryDate": date_str[:10] if len(date_str) >= 10 else None,
                    "supplierName": d.get("supplierName"),
                    "customerId": d.get("customerId"),
                    "article": line.get("article"),
                    "bvlZulassungsnummer": line.get("bvlZulassungsnummer"),
                    "hazardHinweise": line.get("hazardHinweise"),
                    "sdsReference": line.get("sdsReference"),
                    "sachkundeStatus": doc_compliance.get("sachkundeStatus"),
                    "sdsMitgeliefert": doc_compliance.get("sdsMitgeliefert"),
                    "adrPunkte": float(d.get("adrPunkte") or doc_compliance.get("adrPunkte") or 0),
                    "compliant": bool(doc_compliance.get("compliant", False)),
                }
            )
    return rows


def compute_nutrient_stream(deliveries: list[dict], year: int) -> dict:
    by_month: dict[str, dict[str, float]] = defaultdict(lambda: {"deliveries": 0.0, "n_kg": 0.0, "p2o5_kg": 0.0})
    total_n = 0.0
    total_p2o5 = 0.0
    count = 0

    for d in deliveries:
        date_str = str(d.get("date") or "")
        if not date_str.startswith(f"{year}-"):
            continue
        month = date_str[:7]
        n_kg = float(d.get("totalNutrientNKg") or 0.0)
        p2o5_kg = float(d.get("totalNutrientP2o5Kg") or 0.0)

        by_month[month]["deliveries"] += 1
        by_month[month]["n_kg"] += n_kg
        by_month[month]["p2o5_kg"] += p2o5_kg
        total_n += n_kg
        total_p2o5 += p2o5_kg
        count += 1

    return {
        "year": year,
        "deliveryCount": count,
        "totalNutrientNKg": round(total_n, 3),
        "totalNutrientP2o5Kg": round(total_p2o5, 3),
        "byMonth": {
            k: {
                "deliveries": int(v["deliveries"]),
                "n_kg": round(v["n_kg"], 3),
                "p2o5_kg": round(v["p2o5_kg"], 3),
            }
            for k, v in sorted(by_month.items())
        },
    }


def build_lot_trace_report(lot, deliveries: list[dict]) -> dict:
    events = [
        {"type": "charge_created", "date": lot.eingang.isoformat() if lot.eingang else None, "note": "Wareneingang erfasst"},
        {"type": "charge_updated", "date": lot.updated_at.isoformat() if lot.updated_at else None, "note": "Letzte Aktualisierung"},
    ]
    linked_deliveries = []
    for d in deliveries:
        for line in d.get("lines") or []:
            if (line.get("batchNumber") and str(line.get("batchNumber")) == str(lot.chargen_id)) or (
                line.get("articleId") and str(line.get("articleId")) == str(lot.artikel_id)
            ):
                linked_deliveries.append(
                    {
                        "deliveryNumber": d.get("number") or d.get("id"),
                        "deliveryDate": str(d.get("date") or "")[:10],
                        "article": line.get("article"),
                        "quantity": float(line.get("qty") or 0),
                        "customerId": d.get("customerId"),
                    }
                )
                break

    return {
        "lot": {
            "id": lot.id,
            "lotId": lot.chargen_id,
            "article": lot.artikel,
            "articleId": lot.artikel_id,
            "quantity": float(lot.menge or 0),
            "location": lot.lagerort,
            "status": lot.status,
            "qualityStatus": lot.qualitaetsstatus,
            "origin": lot.herkunft,
        },
        "events": events,
        "linkedDeliveries": linked_deliveries,
        "deliveryCount": len(linked_deliveries),
    }
